vm/src/types/structseq.rs

```diff
@@ -0,0 +1,90 @@
+use crate::{
+    builtins::{PyTuple, PyTupleRef, PyTypeRef},
+    pyclass::{PyClassImpl, StaticType},
+    pyobject::PyContext,
+    AsObject, PyObjectRef, PyRef, PyResult, PyValue, VirtualMachine,
+};
+
+#[pyimpl]
+pub trait PyStructSequence: StaticType + PyClassImpl + Sized + 'static {
+    const FIELD_LEN: usize;
+    const FIELD_NAMES: &'static [&'static str];
+
+    fn into_tuple(self, vm: &VirtualMachine) -> PyTuple;
+
+    fn into_struct_sequence(self, vm: &VirtualMachine) -> PyTupleRef {
+        self.into_tuple(vm)
+            .into_ref_with_type(vm, Self::static_type().clone())
+            .unwrap()
+    }
+
+    fn try_elements_from(obj: PyObjectRef, vm: &VirtualMachine) -> PyResult<Vec<PyObjectRef>> {
```

Comment:
Ideally this could return `Box<[PyObjectRef; Self::FIELD_LEN]>` but that's a generic parameter in a const expression :/

Reply:
we usually have solution with generics 😎  https://github.com/RustPython/RustPython/pull/3613/files#diff-c134527406501aa4d7eebd8c9dcd3cfbc8e273cc81cac85e4cba7bf57ddcf707R20-R23